### PR TITLE
A graphical user interface for editing and designing.

### PR DESCRIPTION
**Pull request by Wang-yuanlu**
Hi, Mr Tan. I made some adjustments to the GUI interface and added a black border to each panel to make them look better. Here are the next GUI optimization goals I think I need to do:
1. The tile size in the cgrapanel and mappingpanel can be further adjusted. A 4*4 size model will have some gaps, a 5*5 model size is appropriate, while a 6*6 size is a bit too large.
2. The size of parampanel and testpanel can be further adjusted to make the whole GUI interface look better.
3. The size of DFG can be further adjusted to make the kernelpanel look better.
GUI 4×4
GUI 5×5
GUI 6×6

### CODE REVIEW

**Review by tancheng — COMMENTED** [review]
For the solid borders:

Can you also try to add border into the components inside the `kernel panel` with a screenshot? And would it look better if each button also get the solid border?

For your TODOs:
> The tile size in the cgrapanel and mappingpanel can be further adjusted. A 44 size model will have some gaps, a 55 model size is appropriate, while a 6*6 size is a bit too large.

I think it looks fine for the `5x5` and `6x6`, as we have the horizontal and vertical scrollbars. For the `4x4`, it is better to be adjusted.

I think the issue is the the CGRA panel is not a square and you are determining the size of the tile based on the panel width. Instead, you can determine the tile size based on the `max(panel_height, panel_width)`, and try best to make the `CGRA panel row` and the `kernel panel row` have the same height when the window is popped up.

> The size of parampanel and testpanel can be further adjusted to make the whole GUI interface look better.

Yes, thanks for taking this into consideration, which can be done in the next PR (or this one, you decide).

> The size of DFG can be further adjusted to make the kernelpanel look better.

It looks fine to me now, the only issue is the DFG figure has white background (rather than transparent), and even it is transparent, it cannot fit into the panel. So the solution maybe try to "stretch" the `DFG.png` to fit it into the panel? I believe the tkinter should have such library/interface for that.

You can decide what will be done in each PR, just let me whether existing PR is ready for merge.

File: launchUI.py [hunk 1]
```diff
@@ -1421,8 +1421,8 @@ def create_cgra_pannel(master, rows, columns):
     COLS = columns
     TILE_WIDTH = (GRID_WIDTH + LINK_LENGTH) / COLS - LINK_LENGTH
     TILE_HEIGHT = (GRID_HEIGHT + LINK_LENGTH) / ROWS - LINK_LENGTH
-
-    cgraPannel = tkinter.LabelFrame(master, text='CGRA', bd=BORDER, relief='groove')
+    #Use solid black board to let the pannel look better
```
[the remark this answers is not in the document]
Comment (Wang-yuanlu) [reply launchUI.py:1424]: Okay I will do it.

File: launchUI.py [hunk 2]
```diff
@@ -1692,13 +1696,13 @@ def create_param_pannel(master):
 def create_test_pannel(master):
     dataPannel =tkinter.LabelFrame(master)
     dataPannel.grid(row=0, column=2, rowspan=1, columnspan=1, sticky="nsew")
-    dataPannel.grid_rowconfigure(0, weight=1)
+    # Increase the size of the 'Sverilog' panel
```
[the remark this answers is not in the document]
Comment (Wang-yuanlu) [reply launchUI.py:1699]: Okay I will do it.

File: launchUI.py [hunk 3]
```diff
@@ -1738,53 +1742,53 @@ def create_test_pannel(master):
...
     widgets["reportSPMPowerData"] = reportSPMPowerData
 
-    reportButton.grid(row=0, column=1)
-    reportProgress.grid(row=0, column=0)
+    reportButton.grid(row=0, column=0)
+    reportProgress.grid(row=0, column=1)
 
-    synthesisTimeEntry.grid(row=1, column=0, pady=10)
-    reportTimecostLabel.grid(row=1, column=1, pady=10)
+    synthesisTimeEntry.grid(row=1, column=1, pady=10)
```
[the remark this answers is not in the document]
Comment (Wang-yuanlu) [reply launchUI.py]: Certainly, I realized that 'time entry' used elements from the 'ttk' library, and after unifying them as elements from the 'tkinter' library, they align correctly."

File: launchUI.py [hunk 4]
```diff
@@ -1598,16 +1602,16 @@ def create_param_pannel(master):
     dataMemEntry.insert(0, str(paramCGRA.dataMemSize))
     widgets["dataMemEntry"] = dataMemEntry   
     updateButton = tkinter.Button(paramPannel, text = " Reset ", relief='raised', command = partial(clickReset, master))
-    updateButton.grid(row=2, column=2)
+    updateButton.grid(row=2, column=2, columnspan=2, sticky="nsew")
```
[the remark this answers is not in the document]
Comment (Wang-yuanlu) [reply launchUI.py]: Certainly, I adjusted the relative positions of the buttons to make them align.

**Comment by Wang-yuanlu** [comment]
Hi, Mr Tan. Following your modification suggestions, I made further adjustments and improvements to the GUI interface. I added a black solid board to each component and adjusted the relative sizes of 'parampanel' and 'testpanel', which also made the relative area of 'cgrapanel' more appropriate. Additionally, I optimized the process of generating DFG, allowing the generated DFG graph to be transparently embedded in the corresponding panel.
Here are the next GUI optimization goals I think I need to do:
1. Try to determine the tile size based on the max(panel_height, panel_width).
2. Try to make the CGRA panel row and the kernel panel row have the same height when the window is popped up. However, I believe that this issue is determined by the tkinter library itself. In order to minimize the impact of this issue and ensure that 'cgrapanel' occupies an appropriate area in terms of space, I adjusted the GUI layout to its current style. I believe this can have a certain effect.
GUI 4×4_2

File: launchUI.py [hunk 5]
```diff
@@ -1225,10 +1220,27 @@ def clickShowDFG():
     convertProc = subprocess.Popen([convertCommand, "-u"], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
     (out, err) = convertProc.communicate()
 
+    # Get the size of the whole window
```
Comment (tancheng) [inline launchUI.py:1223]: Lets use `Gets` instead of `Get`.

File: launchUI.py [hunk 6]
```diff
@@ -1225,10 +1220,27 @@ def clickShowDFG():
...
-    PIL_image_small = PIL_image.resize((220, 330), Image.Resampling.LANCZOS)
-    dfgImage = ImageTk.PhotoImage(PIL_image_small)
+    PIL_image_stretched = PIL_image.resize((window_width // 5, window_height // 3), Image.Resampling.BILINEAR)
+    PIL_image_stretched = PIL_image_stretched.convert("RGBA")
+    datas = PIL_image_stretched.getdata()
+
+    new_data = []
+    for item in datas:
+        if item[0] == 255 and item[1] == 255 and item[2] == 255:
+            # Make the image transparent where it is white
```
Comment (tancheng) [inline launchUI.py:1237]: nit: `Makes`.

File: launchUI.py [hunk 7]
```diff
@@ -1311,7 +1323,7 @@ def drawSchedule():
 
     for ii in range(mappingII):
         # draw data memory
-        spmLabel = tkinter.Label(canvas, text="Data\nSPM", fg='black', bg='gray', relief='raised', bd=BORDER)
+        spmLabel = tkinter.Label(canvas, text="Data\nSPM", fg='black', bg='gray', relief='raised', bd=BORDER, highlightbackground="black", highlightthickness=2)
```
Comment (tancheng) [inline launchUI.py:1326]: Would it look better if we make the border a little thinner? Say `thickness = 1`? WDYT?

Comment (tancheng) [reply launchUI.py:1326]: And I think the `thickness` could also be a global constant variable, right?

**Review by tancheng — COMMENTED** [review]
Hi Yuanlu, thanks for the quick fix.

 - You can actually reply reviewer's comment at the corresponding place after you made (commit) the change, so it would be more clear to the reviewer what is changed.
 - Do you think it would look better if we also add border to the `label`s?
 - Can we avoid adding border to the `panel`? The panel itself has some border with name embedded on top, adding border to panel makes it look weird.
 - How did you make the `DFG.png` be transparent? It is done outside the GUI using other tools? If yes, the issue would still be there, e.g., when you enlarge/shrink the window, the showing `DFG` won't be scaled up/down.

File: launchUI.py (hunk 5 above)
Comment (Wang-yuanlu) [reply launchUI.py:1223]: Okay I will do it.

File: launchUI.py (hunk 6 above)
Comment (Wang-yuanlu) [reply launchUI.py:1237]: Okay I will do it.

File: launchUI.py (hunk 7 above)
Comment (Wang-yuanlu) [reply launchUI.py:1326]: Certainly, I set the thickness as a global variable and I set the thickness to 1. It seems that let thickness = 1 will make the graphical user interface more clear and concise.

**Comment by Wang-yuanlu** [comment]
Hi, Mr Tan. Following your modification suggestions, I made further adjustments and improvements to the GUI interface.
I change the thickness of the solid borad to make the GUI more clear and concise. I used functions from PIL library which can be used to create transparent backgrounds or overlay the image on top of other content and then optimizing the DFG.
GUI thickness=1

**Comment by Wang-yuanlu** [comment]
Hi, Mr Tan. Following your modification suggestions, I made further adjustments and improvements to the GUI interface.
I change the thickness of the solid borad to make the GUI more clear and concise. I used functions from PIL library which can be used to create transparent backgrounds or overlay the image on top of other content and then optimizing the DFG.
The results are as follows. Have a nice day!
GUI thickness=1

File: launchUI.py [hunk 8]
```diff
@@ -30,6 +30,7 @@
 MEM_WIDTH = 50
 CONFIG_MEM_SIZE = 8
 DATA_MEM_SIZE = 4
+thickness = 1
```
Comment (tancheng) [inline launchUI.py:33]: Plz rename this as `HIGHLIGHT_THICKNESS`.

**Comment by tancheng** [comment]
Hi Yuanlu, thanks for the update.
 - Is it possible to add a tiny interval between labels/fields. They are too closed to each other. Say,
   - `Time cost:` and its corresponding field in "report area/power".
   - `Time cost:`'s field is too closed to the boundary of the entire panel.
   - Also the `Application:`'s field and the `Compile app`.
 - Now the `DFG.png` can be stretched, right? Can you plz attach an enlarged window to show case the stretched `DFG.png`? I am thinking maybe the white background is more suitable than the transparent version.

File: launchUI.py (hunk 8 above)
Comment (Wang-yuanlu) [reply launchUI.py]: okay I will do it.

**Comment by Wang-yuanlu** [comment]
Hi Mr.Tan. I have further optimized the positioning of the components in the param module, and adjusted the spacing between the components in the GUI as well as the spacing between the components and the boundaries. Regarding the DFG diagram part, I have adjusted it back to a white background.

**Comment by tancheng** [comment]
Hi Yuanlu, thanks for the update, can u also share the screenshot?

**Comment by Wang-yuanlu** [comment]
> Hi Yuanlu, thanks for the update, can u also share the screenshot?

Okay. Here is the screenshot:
GUI now

**Comment by tancheng** [comment]
Cool, can u stretch out the window and paste the screenshot again? I wanna take a look at the stretched out DFG graph. Thanks!

**Comment by Wang-yuanlu** [comment]
> Cool, can u stretch out the window and paste the screenshot again? I wanna take a look at the stretched out DFG graph. Thanks!

Okay. 
GUI now 2 0
However, the stretching effect of the DFG diagram is not very pronounced.

**Comment by tancheng** [comment]
Do u think it is possible to make the `DFG.png` automatically align with the panel size (we can sacrifice the resolution)?

**Comment by Wang-yuanlu** [comment]
> Do u think it is possible to make the `DFG.png` automatically align with the panel size (we can sacrifice the resolution)?

I think that I am currently unable to adjust the adaptation of the DFG graph. It is embedded as a fixed-size component within the kernel panel and I have tried to add scrollbars for it, but the corresponding panel structure cannot accommodate the scrollbars.

**Comment by tancheng** [comment]
No problem. Thanks! Let me merge the PR.

**Pull request merged**